xrdp/xrdp_egfx.c

```diff
@@ -0,0 +1,1021 @@
+    LOG(LOG_LEVEL_DEBUG, "xrdp_egfx_send_create_surface:"
+        " xrdp_egfx_send_data error %d", error);
+    free_stream(s);
+    return error;
+}
+
+/******************************************************************************/
+int
+xrdp_egfx_send_delete_surface(struct xrdp_egfx *egfx, int surface_id)
+{
+    int error;
+    int bytes;
+    struct stream *s;
+    char *holdp;
+
+    LOG(LOG_LEVEL_TRACE, "xrdp_egfx_send_delete_surface:");
+    make_stream(s);
+    init_stream(s, 8192);
+    /* RDP_SEGMENTED_DATA */
+    out_uint8(s, 0xE0); /* descriptor = SINGLE */
+    /* RDP8_BULK_ENCODED_DATA */
+    out_uint8(s, 0x04); /* header = PACKET_COMPR_TYPE_RDP8 */
+    /* RDPGFX_HEADER */
+    out_uint16_le(s, XR_RDPGFX_CMDID_DELETESURFACE); /* cmdId */
+    out_uint16_le(s, 0); /* flags = 0 */
+    holdp = s->p;
+    out_uint8s(s, 4); /* pduLength, set later */
+    out_uint16_le(s, surface_id);
+    s_mark_end(s);
+    bytes = (int) ((s->end - holdp) + 4);
+    s->p = holdp;
+    out_uint32_le(s, bytes);
+    bytes = (int) (s->end - s->data);
+    error = xrdp_egfx_send_data(egfx, s->data, bytes);
+    LOG(LOG_LEVEL_DEBUG, "xrdp_egfx_send_delete_surface:"
+        " xrdp_egfx_send_data error %d", error);
+    free_stream(s);
+    return error;
+}
+
+/******************************************************************************/
+int
+xrdp_egfx_send_map_surface(struct xrdp_egfx *egfx, int surface_id,
+                           int x, int y)
+{
+    int error;
+    int bytes;
+    struct stream *s;
+    char *holdp;
+
+    LOG(LOG_LEVEL_TRACE, "xrdp_egfx_send_map_surface:");
+    make_stream(s);
+    init_stream(s, 8192);
+    /* RDP_SEGMENTED_DATA */
+    out_uint8(s, 0xE0); /* descriptor = SINGLE */
+    /* RDP8_BULK_ENCODED_DATA */
+    out_uint8(s, 0x04); /* header = PACKET_COMPR_TYPE_RDP8 */
+    /* RDPGFX_HEADER */
+    out_uint16_le(s, XR_RDPGFX_CMDID_MAPSURFACETOOUTPUT); /* cmdId */
+    out_uint16_le(s, 0); /* flags = 0 */
+    holdp = s->p;
+    out_uint8s(s, 4); /* pduLength, set later */
+    out_uint16_le(s, surface_id);
+    out_uint16_le(s, 0);
+    out_uint32_le(s, x);
+    out_uint32_le(s, y);
+    s_mark_end(s);
+    bytes = (int) ((s->end - holdp) + 4);
+    s->p = holdp;
+    out_uint32_le(s, bytes);
+    bytes = (int) (s->end - s->data);
+    error = xrdp_egfx_send_data(egfx, s->data, bytes);
+    LOG(LOG_LEVEL_DEBUG, "xrdp_egfx_send_map_surface:"
+        " xrdp_egfx_send_data error %d", error);
+    free_stream(s);
+    return error;
+}
+
+/******************************************************************************/
+int
+xrdp_egfx_send_fill_surface(struct xrdp_egfx *egfx, int surface_id,
+                            int fill_color, int num_rects,
+                            const struct xrdp_egfx_rect *rects)
+{
+    int error;
+    int bytes;
+    int index;
+    struct stream *s;
+    char *holdp;
+
+    LOG(LOG_LEVEL_TRACE, "xrdp_egfx_send_fill_surface:");
+    make_stream(s);
+    init_stream(s, 8192);
```

Comment:
Is this guaranteed to be big enough, or should you scale it depending on `num_rects`? Somehing like `init_stream(s, 1024 + num_rects * 8);`

Reply:
yes, unlikely but yes.

Reply:
@matt335672 I believe this is resolved.